src/molden_utils.F

```diff
@@ -131,30 +132,35 @@ SUBROUTINE write_mos_molden(mos, qs_kind_set, particle_set, print_section)
+                                         npgf=npgf, &
                                          nshell=nshell, &
-                                         l=l)
-
-                  DO iset = 1, orb_basis_set%nset
-                     DO ishell = 1, orb_basis_set%nshell(iset)
-                        IF (orb_basis_set%l(ishell, iset) == 0) angmom = 's'
-                        IF (orb_basis_set%l(ishell, iset) == 1) angmom = 'p'
-                        IF (orb_basis_set%l(ishell, iset) == 2) angmom = 'd'
-                        IF (orb_basis_set%l(ishell, iset) == 3) angmom = 'f'
-                        IF (orb_basis_set%l(ishell, iset) == 4) angmom = 'g'
-                        WRITE (UNIT=iw, &
-                               FMT="(T25,A2,4X,I4,4X,F4.2)") &
-                           angmom, &
-                           orb_basis_set%npgf(iset), 1.0
-                        ! MOLDEN expects the contaction coefficient of spherical NOT CARTESIAN NORMALISED
-                        ! functions. So we undo the normalisation factors included in the gccs
-                        ! Reverse engineered from basis_set_types, normalise_gcc_orb
-                        prefac = 2_dp**orb_basis_set%l(ishell, iset)*(2/pi)**0.75_dp
-                        expzet = 0.25_dp*(2*orb_basis_set%l(ishell, iset) + 3.0_dp)
-                        WRITE (UNIT=iw, &
-                               FMT="((T51,2F15.6))") &
-                           (orb_basis_set%zet(ipgf, iset), &
-                            orb_basis_set%gcc(ipgf, ishell, iset)/(prefac*orb_basis_set%zet(ipgf, iset)**expzet), &
-                            ipgf=1, orb_basis_set%npgf(iset))
+                                         l=l, &
+                                         zet=zet, &
+                                         gcc=gcc)
+
+                  DO iset = 1, nset
+                     DO ishell = 1, nshell(iset)
+                        lshell = l(ishell, iset)
+                        IF (lshell <= molden_lmax) THEN
+                           WRITE (UNIT=iw, &
+                                  FMT="(T25,A2,4X,I4,4X,F4.2)") &
+                              angmom(lshell + 1:lshell + 1), npgf(iset), 1.0
+                           ! MOLDEN expects the contraction coefficient of spherical NOT CARTESIAN NORMALISED
+                           ! functions. So we undo the normalisation factors included in the gccs
+                           ! Reverse engineered from basis_set_types, normalise_gcc_orb
+                           prefac = 2_dp**lshell*(2/pi)**0.75_dp
+                           expzet = 0.25_dp*(2*lshell + 3.0_dp)
+                           WRITE (UNIT=iw, &
+                                  FMT="((T51,2F15.6))") &
+                              (zet(ipgf, iset), gcc(ipgf, ishell, iset)/(prefac*zet(ipgf, iset)**expzet), &
+                               ipgf=1, npgf(iset))
+                        ELSE
+                           IF (print_warn) THEN
```

Comment:
`print_warn` is still uninitialized at this point! (thanks to Alfio, Cray Compiler)

Reply:
Thank you for pointing it out!